prefect_saturn/core.py

```diff
@@ -136,91 +148,131 @@ def register_flow_with_saturn(self, flow: Flow) -> bool:
-        The image name is generated in Saturn.
+    def register_flow_with_saturn(
+        self, flow: Flow, dask_cluster_kwargs=None, dask_adapt_kwargs=None
+    ) -> Flow:
         """
-        storage_details = self.storage_details
-        image_tag = str(uuid.uuid4())
-        # NOTE: SATURN_TOKEN and BASE_URL have to be set to be able
-        #       to load the flow. Those variables will be overridden by
-        #       Kubernetes in all the places where it matters, and values
-        #       set in the image are overridden by those from kubernetes
-        #       https://kubernetes.io/docs/tasks/inject-data-application/define-environment-variable-container
-        flow.storage = Docker(
-            base_image=self._saturn_base_image,
-            image_name=storage_details["image_name"],
-            image_tag=image_tag,
-            registry_url=storage_details["registry_url"],
-            prefect_directory="/tmp",
-            env_vars={"SATURN_TOKEN": "placeholder-token", "BASE_URL": "placeholder-url"},
-            python_dependencies=["kubernetes"],
-            ignore_healthchecks=True,
+        Given a flow, set up all the details needed to run it on
+        a Saturn Dask cluster.
+
+        This method modifies the following components of ``Flow`` objects
+        passed to it.
+
+        * ``.storage``: a ``Webhook`` storage instance is added
+        * ``.environment``: a ``KubernetesJobEnvironment`` with a ``DaskExecutor``
+            is added. This environment will use the same image as the notebook
+            from which this code is run.
+        """
+        self._set_flow_metadata(flow)
+
+        storage = self._get_storage()
+        flow.storage = storage
+
+        environment = self._get_environment(
+            cluster_kwargs=dask_cluster_kwargs, adapt_kwargs=dask_adapt_kwargs
         )
+        flow.environment = environment
+
+        flow.storage.add_flow(flow)
+
+        # tell Saturn to store the flow
+        self._store_flow(flow=flow)
         return flow
 
-    def build_storage(self, flow: Flow) -> Response:
+    def _get_storage(self) -> Webhook:
         """
-        Actually build and push the storage
+        Create a `Webhook` storage object with Saturn-y details.
         """
-        if self._saturn_flow_id is None:
-            raise RuntimeError(Errors.NOT_REGISTERED)
+        storage = Webhook(
+            build_request_kwargs={
+                "url": "${BASE_URL}api/prefect_cloud/flows/" + self.flow_id + "/flow_content",
+                "headers": {
+                    "Content-Type": "application/octet-stream",
+                    "Authorization": "token ${SATURN_TOKEN}",
+                },
+            },
+            build_request_http_method="POST",
+            get_flow_request_kwargs={
+                "url": "${BASE_URL}api/prefect_cloud/flows/"
+                + self.flow_id
+                + "/flow_content?flow_version_id="
```

Comment:
Is there a reason we are using url query string here versus another endpoint that includes the version id?

Reply:
so, the way I'm thinking about this, `/api/prefect_cloud/flows/{flow_id}/flow_content` references one flow "resource", and then `{flow_version_id}` is a filter on the list of all versions of that resource.

I thought that was more natural than having the idea of a `flow_version` resource, esp. since flow versions aren't tracked in the app's database.

If you have a strong opinion I'll change it.

Reply:
I think that is a totally valid approach, but it created a bit of confusion for me.

If the GET and POST require a `flow_version_id` I would expect it to be part of the URL and not a required query string since query strings do not have to be included to fetch a URL.
If `...{flow_id}/flow_content` worked without `flow_version_id` then I would expect it to be a query string. It could work that way, perhaps that endpoint gives all versions, and if you supply the query string it gives back only one?

Right now based on how these endpoints are written I would expect it to look like: `/api/prefect_cloud/flows/{flow_id}/{flow_version}/content` or similar.

Reply:
> If ...{flow_id}/flow_content worked without flow_version_id then I would expect it to be a query string. It could work that way, perhaps that endpoint gives all versions, and if you supply the query string it gives back only one?

I think one crucial difference here is that the API is all internal, and we're only building exactly the endpoints needed by Saturn stuff. If we were maintaining a publicly-accessible API we'd implement every endpoint to be sure there are no orphaned paths, but that would be wasted effort as long as our API is only used internal to Saturn's product.

Anyway, I'm ok with `/api/prefect_cloud/flows/{flow_id}/{flow_version}/content`. I'll make that change.

Reply:
I definitely agree with you, and I think it would be wasted effort to make our internal API be designed and maintained as if public-accessible. I do think having the URLs be as self-explanatory as possible is important though which is why I suggested the change here.
With the change I wouldn't need to look up our API endpoint to understand a bit better how this works. Hopefully that is true for others.

Reply:
ok I just made this change here and in the corresponding Saturn back-end PR!